wdl-format/src/comments.rs

```diff
@@ -0,0 +1,192 @@
+//! Format comments in a WDL file.
+//!
+//! All comments will be treated as either "preceding" or "inline" comments.
+//! Every comment will "belong" or "be owned" by a specific element in the
+//! syntax tree. If that element is moved from one place to another, the
+//! comment will move with it. Only syntax elements that are either the first
+//! element of a line or the last element of a line can own comments. Elements
+//! may span multiple lines, only the beginning (in the case of preceding
+//! comments) or the end (in the case of inline comments) of the element
+//! are considered.
+//!
+//! A preceding comment is a comment that appears on a line before an element.
+//! There may be any number of preceding comments and they may be separated
+//! by any number of blank lines. Internal blank lines (blanks _between_
+//! comments) will be included in the formatted output, but blank lines before
+//! or after all comments will not be includes. Multiple internal blank lines in
+//! a row will be consolidated to one. Preceding comments should be indented to
+//! the same level as the element which they belong to.
+//!
+//! An inline comment is a comment that appears on the same line as an element,
+//! if and only if that element is the last element of its line. Inline comments
+//! should always appear two spaces after the element they are commenting on.
+
+use wdl_ast::AstToken;
+use wdl_ast::Comment;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+
+use super::Formattable;
+use super::Formatter;
+use super::NEWLINE;
+
+/// Inline comment space constant used for formatting. Inline comments should
+/// start two spaces after the end of the element they are commenting on.
+pub const INLINE_COMMENT_SPACE: &str = "  ";
+
+impl Formattable for Comment {
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        _formatter: &mut Formatter,
+    ) -> std::fmt::Result {
+        let comment = self.as_str().trim();
+        write!(writer, "{}{}", comment, NEWLINE)
+    }
+}
+
+/// Format comments that precede a node.
+///
+/// This function assumes we are _not_ in the preamble, and thus any
+/// double-pound-sign comments should be converted to single-pound-sign
+/// comments.
+pub fn format_preceding_comments<T: std::fmt::Write>(
+    element: &SyntaxElement,
+    writer: &mut T,
+    formatter: &mut Formatter,
+    would_be_interrupting: bool,
+) -> std::fmt::Result {
+    // This walks _backwards_ through the syntax tree to find comments
+    // so we must collect them in a vector and later reverse them to get them in the
```

Comment:
Rather than doing this, you should just use `VecDeque::push_front`. Because `VecDeque` uses a ring buffer internally, that's a constant time operations with no need to reverse the elements.

Reply:
I didn't put this comment everywhere, but, in general, you should use that instead of gathering the elements in reverse and then reversing them.